crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+                Some(other) => ty.is_subtype_of(other),
+                None => false,
+            })
+            && other
+                .exports
+                .iter()
+                .all(|(k, other)| match self.exports.get(k) {
+                    Some(ty) => ty.is_subtype_of(other),
+                    None => false,
+                })
+    }
+}
+
+#[derive(Clone)]
+struct ModuleInstanceType {
+    exports: Rc<HashMap<String, EntityType>>,
+}
+
+#[derive(Clone)]
+struct ComponentType {
+    imports: Rc<HashMap<String, ComponentEntityType>>,
+    exports: Rc<HashMap<String, ComponentEntityType>>,
+}
+
+impl ComponentType {
+    fn is_subtype_of(&self, other: &Self) -> bool {
+        self.imports
+            .iter()
+            .all(|(k, ty)| match other.imports.get(k) {
+                Some(other) => ty.is_subtype_of(other),
+                None => false,
+            })
+            && other
+                .exports
+                .iter()
+                .all(|(k, other)| match self.exports.get(k) {
+                    Some(ty) => ty.is_subtype_of(other),
+                    None => false,
+                })
+    }
+}
+
+#[derive(Clone)]
+struct InstanceType {
+    exports: Rc<HashMap<String, ComponentEntityType>>,
+}
+
+impl InstanceType {
+    fn is_subtype_of(&self, other: &Self) -> bool {
+        other
+            .exports
+            .iter()
+            .all(|(k, other)| match self.exports.get(k) {
+                Some(ty) => ty.is_subtype_of(other),
+                None => false,
+            })
+    }
+}
+
+#[derive(Clone, Eq)]
+struct ComponentFuncType {
+    params: Rc<[(String, InterfaceType)]>,
+    result: InterfaceType,
+    core_type: Arc<FuncType>,
+}
+
+impl PartialEq for ComponentFuncType {
+    fn eq(&self, other: &Self) -> bool {
+        self.params
+            .iter()
+            .map(|(_, ty)| ty)
+            .eq(other.params.iter().map(|(_, ty)| ty))
+            && self.result.eq(&other.result)
+    }
+}
+
+#[derive(Clone, Eq)]
+enum CompoundType {
+    Record(Box<[InterfaceType]>),
+    Variant(Box<[InterfaceType]>),
+    List(InterfaceType),
+    Tuple(Box<[InterfaceType]>),
+    Flags(usize),
+    Enum(usize),
+    Union(Box<[InterfaceType]>),
+    Optional(InterfaceType),
+    Expected(InterfaceType, InterfaceType),
+}
+
+impl PartialEq for CompoundType {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Record(l), Self::Record(r))
+            | (Self::Variant(l), Self::Variant(r))
+            | (Self::Tuple(l), Self::Tuple(r))
+            | (Self::Union(l), Self::Union(r)) => l == r,
+
+            (Self::List(l), Self::List(r)) => l == r,
+            (Self::Flags(l), Self::Flags(r)) | (Self::Enum(l), Self::Enum(r)) => l == r,
+            (Self::Optional(l), Self::Optional(r)) => l == r,
+            (Self::Expected(lo, le), Self::Expected(ro, re)) => lo == ro && le == re,
+            _ => false,
+        }
+    }
+}
```

Comment:
Hm ok this raises some questions for me, probably all spec related:

* I'm curious why equality is necessary here vs subtyping? I thought we were going to have something along the lines of a struct is a subtype of another if it has less fields (or something like that). Basically I thought there was going to be some story of adding fields backwards-compatibly (or removing them? I forget) but if this strict equality is used then I don't think that's the case. (although again I've only read thus far, I may be missing something later).

* This doesn't reflect what I thought flags/enum/union/etc special cases were going to be. Originally those started out as shorthands for other representations, but given this interpretation they're entirely distinct. For example an enum and a variant with the same names would be considered distinct types, but I thought they were supposed to be equal and only affect how languages generate bindings? (now I'm basically not sure why there's specialized types)

* For variants and records I thought that the order of variants/fields didn't matter so long as the names lined up? I thought that we weren't going to require a strict ordering amongst modules allowing everyone to flexibly define fields and consume them in the order the language embedding requires.

Reply:
I agree that it is a little weird to have equality here when it's a subtype check. The spec hasn't fleshed out the rules for instance/module/component subtyping where interface types are involved (yet), but I personally don't see how we can accept anything but equality for the interface types because altering fields of a struct (even just appending one) or the cases of a variant alters the canonical ABI signature (i.e. the underlying core function type might not match between the import and export).

The module subtyping rules in the former module linking proposal used strict equality of function types too, although obviously that proposal wasn't also complicated with supported interface types at the same time. I definitely think this is one of the areas we will just need to fix later in the validator once the spec is filled in more.

Regarding the variants of `CompoundType`, I agree that there seems to be unnecessary duplication in the type representation. I'll add that to the list of things to bring up with the spec.

Regarding field/case ordering in records/variants, I think there needs to be a _definitive_ order somewhere, ideally in the representation in the component itself, such that we know what the canonical ABI representation should be. The language bindings should be free to use any order they wish such that when converting to and from the canonical ABI the _definitive_ order is respected.

Reply:
I mean, we _could_ have the adapter functions themselves be smart enough to know the caller and callee's layouts of the same interface type (provided it passes a subtype check) and do the magic to order things appropriately for the canonical ABI, but that seems like a bit of a burden to put on implementer, perhaps?

Still, as you pointed out, that might be needed _anyway_ if component interfaces can be backwards compatible in these ways (e.g. adding an optional field to a record).

I think we should land this PR with not-yet-correct subtyping checks and update when there's more guidance from the spec.

Reply:
Yeah landing for now seems fine, but these are what I thought were pretty important parts of the interface types vision which we'll need to address. The theory was always that modules don't have to agree on ABIs, only on interface types, which should provide some measure of flexibility. The canonical ABI is intended to be a way to synthesize lower/lift functions but other than that shouldn't affect the actual interface types semantics at the spec level.

Reply:
I've pushed up subtype checking for the interface types.